### DISCUSSION TITLE
[IntelliSense] Add support for Laravel Blade @class directive and for Laravel Blade component attributes functions

### DISCUSSION THREAD

**Discussion by alexmandrikdev** [tailwindlabs/tailwindcss#7046] [Ideas]
Please make this extension work with [@class directive](https://laravel.com/docs/8.x/blade#conditional-classes) and with [attributes functions](https://laravel.com/docs/8.x/blade#conditionally-merge-classes).

**Comment by ph7jack** [comment]
UP

**Reply by ph7jack** [reply]
Would be great to have class syntax in the laravel blade **@.class** directive
![image](https://user-images.githubusercontent.com/38913462/182651050-8271adf2-ba25-4162-b61e-939655e4387e.png)
![image](https://user-images.githubusercontent.com/38913462/182651090-1ace2f81-ac4a-4d2b-a67e-7206ed8b6bd2.png)

**Reply by TimvdHorst** [reply]
@alexmandrikdev maybe change the title and start it with [IntelliSense] to make it more clearer what it is about? Would also love this to work in the IntelliSense VSCode extension.

**Comment by czernika** [comment]
This is not entirely my solution but I don't remember where I've found original article

You can add custom class-pattern in extension settings. Add these lines into `settings.json`

```json
"tailwindCSS.experimental.classRegex": [
     ["@class\\(([^]*)\\)", "'([^']*)'"]
 ]
``` 

It should work (custom values and variants included)

![Снимок экрана от 2022-10-13 14-38-43](https://user-images.githubusercontent.com/86893348/195587176-c3f6bc5d-d753-4653-80cd-f272428c5de7.png)

**Reply by alexmandrikdev** [reply]
Thank you for this solution, @czernika!

**Reply by alexmandrikdev** [reply]
Do you know why it doesn't work with classes using `,` characters?

For example: `shadow-[0px_2px_20px_rgba(0,0,0,0.4)]`

**Reply by czernika** [reply]
Ah yes, good catch. I've edited answer - instead of `"@class\\(([^)]*)\\)"` it should be `"@class\\(([^]*)\\)` (one extra bracket was removed)
![Снимок экрана от 2022-10-14 10-50-35](https://user-images.githubusercontent.com/86893348/195793077-1d6417cd-438f-4a55-844f-f806d6c8faa6.png)

**Reply by alexmandrikdev** [reply]
Thank you!

**Reply by alexmandrikdev** [reply]
If you remove the `@`, it became works with the `$attributes->class()` method too.

``` json
"tailwindCSS.experimental.classRegex": [
     ["class\\(([^]*)\\)", "'([^']*)'"]
]
```

![image](https://user-images.githubusercontent.com/36386967/195837910-c12e5935-bbfe-4a9c-a752-4c506e3e969a.png)

**Reply by czernika** [reply]
Thanks didn't know about that! You may put `?` sign after `@` (like `@` may or may not be present) so regex should be `["@?class\\(([^]*)\\)", "'([^']*)'"]` - this way both `@class` and `$attributes->class()` should work as expected

**Reply by tobia** [reply]
Thank you. Here is an updated version:
```json
"tailwindCSS.experimental.classRegex": [
    ["class\\((.*)\\)", "'([^']*)'"],
    ["class\\((.*)\\)", "\"([^\"]*)\""],
]
```
This recognizes both single-quoted and double-quoted PHP strings. But it's still incomplete, because it will break when a single-quoted string contains an escaped single-quote, and other edge cases. 

Unfortunately a full solution is not possible without a full PHP parser, or at least a Regex engine with support for recursive patterns (like `(?1)` in Perl or PCRE) which AFAIK is not present in Node's default Regex.
